File: cpp/templates/strings/HashedString.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
class HashedString {
  private:
	static const ll M = 1e9 + 9, B = 9973;
	static vector<ll> pow;
	vector<ll> p_hash;
  public:
	HashedString(const string &s) : p_hash(s.size() + 1) {
    if (pow.size() == 0) pow.push_back(1);
		while (pow.size() < s.size()) { 
      pow.push_back((pow.back() * B) % M); 
    }
		p_hash[0] = 0;
		for (int i = 0; i < s.size(); i++) {
			p_hash[i + 1] = ((p_hash[i] * B) % M + s[i]) % M;
		}
	}
	ll getHash(int start, int end) {
		ll raw_val = (p_hash[end + 1] - (p_hash[start] * pow[end - start + 1]));
		return (raw_val % M + M) % M;
	}
};

File: python/strings/HashedString.py
class HashedString:
  M, B, pow = 1e9 + 9, 9973, [1]
  
  def __init__(self, s):
    self.p_hash = [0] * (len(s) + 1)
    while len(self.pow) <= len(s):
      self.pow.append((self.pow[-1] * self.B) % self.M)
    
    for i in range(len(s)):
      self.p_hash[i + 1] = (self.p_hash[i] * self.B + ord(s[i])) % self.M

  def get_hash(self, start, end):
    raw_val = (self.p_hash[end + 1] - (self.p_hash[start] * self.pow[end - start + 1])) % self.M
    return (raw_val + self.M) % self.M
